Align columns in Targets view

applegrain opened the pull request:
Fixes #3373 

Small PR to align the columns in the Target view. 

I ended up refactoring the nested tables, wasn't entirely clear to me why that was needed. A consequence of that refactor was a slight style change in terms of the table headers. I also added "show more/show less" text to communicate that you can expand and collapse the tables.

![](http://g.recordit.co/w8RElnXgb0.gif)

grobie commented:
Hm, defining fixed ratios will be difficult. What happens if a column becomes larger than its percentage? Endpoint, labels and Error columns can all becomes arbitrary large (Endpoint less likely so). Ednpoint, State and Last Scrape could just be non-wrapping and as small as possible.

Before 245b8a0b379e7c25a34e37315b06c81c755814a0 we repeated thead and tbody groups, and the browser aligned the able automatically for us choosing the best width. I just read that it's actually illegal to render a thead after a tbody element. Any other way to letting the browser do its job without the need to specify the percentage on our own?

applegrain commented:
Hey! I've pushed some more changes to this PR. To avoid breaking HTML law, I refactored away the nested tables. I kept the ratios to ensure consistent cell width across all separate tables. If the values exceed the cell width, the cell height will expand and make room for the information.

@grobie also mentioned on irc that it would be useful to be able to filter targets based on whether or not they're healthy, and I've added that functionality as well. Let me know what you think!

![targetview-short](https://user-images.githubusercontent.com/10555255/32747338-b86b6bda-c875-11e7-87ca-afdc7fcf12f1.gif)

grobie commented:
Thanks so much @applegrain! I'll test this on a server with many targets tomorrow!

applegrain commented:
Perfect! I look forward to your bug report 😉

juliusv commented:
Awesome, and we finally have a frontend dev! :)

Codacy is complaining about single quotes in JS. It's not like we ever came up with this rule or adhere to it everywhere in other places in the file, but we should probably convert everything to double quotes in an unrelated PR. But just the new lines for now...

grobie commented:
We completely disabled codacy today, as it caused many false alerts (for go
at least). We could just ignore these errors.

On Tue, Nov 14, 2017 at 12:28 AM Julius Volz <notifications@github.com>
wrote:

> Awesome, and we finally have a frontend dev! :)
>
> Codacy is complaining about single quotes in JS. It's not like we ever
> came up with this rule or adhere to it everywhere in other places in the
> file, but we should probably convert everything to double quotes in an
> unrelated PR. But just the new lines for now...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/3445#issuecomment-344094158>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaDlFwRQKfof6bGyMPqULv8h7fVtCks5s2NCmgaJpZM4QXLIu>
> .
>

juliusv commented:
Ah, great as well.

applegrain commented:
Changed the single quotes to double quotes, regardless of what the linter says it's nice to be consistent..

juliusv commented:
@applegrain Just trying this out locally now.

A couple of nits:

- On Chrome+Linux, the buttons look like this: ![buttons](https://user-images.githubusercontent.com/538008/32790083-bb330204-c95d-11e7-8f65-cb6cf345fd75.png) Would it be possible to style them similarly to the UI elements on the /graph page?
- The text "View Unhealthy Targets" confused me a bit because it showed also healthy targets, but only in jobs that also had unhealthy targets. Maybe "View jobs with unhealthy targets"?
- The buttons don't get highlighted to show what is currently selected. I am wondering whether two buttons is the way to go for a toggle. How about making it a checkbox that is styled like the "stacked" button on [http://demo.robustperception.io:9090/graph?g0.range_input=1h&g0.expr=&g0.tab=0](http://demo.robustperception.io:9090/graph?g0.range_input=1h&g0.expr=&g0.tab=0). Then the text of the checkbox button could simply be "Only jobs with unhealthy targets". Perhaps even shorter, "Only unhealthy jobs".

applegrain commented:
Thanks for the feedback @juliusv! I'll work on getting those changes in now:  
- I'll make the styling consistent with the elements on the /graph page
- I like your idea of using a checkbox as a toggle, there's no need for having two buttons

applegrain commented:
Latest changes:
![latest-trimmed](https://user-images.githubusercontent.com/10555255/32797691-c92c5bae-c92f-11e7-8eb1-6b170c913b95.gif)

juliusv commented:
@applegrain I really like it now from what it looks like. Will also look at the code now.

Could you run `make assets` again to make the tests pass?

juliusv commented on web/ui/static/css/targets.css:
```diff
@@ -1,11 +1,52 @@
+    color: rgb(242, 65, 65);
+}
+
+.filters button.unhealthy-targets {
```
This checkbox button business should probably be refactored into a shared CSS file, but I think that doesn't need to happen in this PR.

applegrain replied:
I <3 refactoring, so happy to do that after this is done!

juliusv replied:
Cool, same applies to the related JS :)

juliusv commented on web/ui/static/js/targets.js:
```diff
@@ -1,38 +1,62 @@
-    var icon = $(obj).find("i");
-    if (icon.length === 0 ) {
-        return;
+function toggle(button, state){
```
How about `toggleJobTable()` to make the action clearer (since there's different kinds of toggling happening on the page)?

juliusv commented on web/ui/static/js/targets.js:
```diff
@@ -1,38 +1,62 @@
+function toggle(button, state){
+  if (button.length === 0) { return; }
+
+  if (state) {
```
To clarify that `state` is a boolean variable (vs. e.g. some enum) and what its `true` and `false` values mean, how about calling it `isCollapsed`?

juliusv commented:
Looks good to me now, thanks! I didn't look tooo close at the JS+HTML because that's not that much of my thing, but at least it all seems to work for me :)

@grobie What do you think?

applegrain commented:
ping @grobie, did you have a chance to look at the latest changes? Let me know what you think!

Gouthamve commented:
Can confirm it works for me 👍

juliusv commented:
If there's no response from @grobie by tomorrow, I'll just merge it. @applegrain Could you rebase + `make assets` one more time to get rid of the merge conflict?

applegrain commented:
Done ✔️ 
cc @juliusv

juliusv commented:
@applegrain Thanks a bunch!
